core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -5120,12 +5112,30 @@ else if (!(isEffectivelyConstantExpression(node) && n > 1)) {
                 } else {
                     emitCursors(qc.getAst(), cursorModel, null, translatingModel, baseModel, sqlExecutionContext, sqlParserCallback);
                     if (useGroupByModel) {
+                        // exclude constant columns from group-by, for example:
+                        // select 1, id, sum(x) from ...
+                        // keying map on constant '1' is unnecessary; this column can be selected
+                        // after the group-by in the "outerVirtualModel"
                         if (isEffectivelyConstantExpression(qc.getAst())) {
                             outerVirtualIsSelectChoose = false;
                             outerVirtualModel.addBottomUpColumn(qc);
                             distinctModel.addBottomUpColumn(qc);
                             continue;
                         }
+
+                        // sample-by queries will still require innerVirtualModel
```

Comment:
nit: the below `if (sampleBy == null)` only applies to group-by, so this comment is a bit confusing.

Reply:
"sample by" does not have own model, from this code perspective "sample by" is a specialized group-by. The comment is deliberate actually. I accept it may be more detailed. Let me know if this is the case.

This is how we generate sample by:

![image](https://github.com/questdb/questdb/assets/7276403/09613598-83d5-4f3f-a79b-c620dd06efa8)

Reply:
> "sample by" does not have own model, from this code perspective "sample by" is a specialized group-by.

Yes, I know that. The comment feels misleading since it's placed before a group-by specific block of code, so it's worth expanding it.